From 8d5c938e4313d9fe96a53e685f7a41b06f8d4f71 Mon Sep 17 00:00:00 2001
From: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
Date: Fri, 21 Jun 2024 11:42:02 -0700
Subject: [PATCH] QSDK: wifi: ath12k: Fix incorrect dereferencing driver's VIF
 context

Currently when ECM issues a callbacks to ath12k driver for
fetching DS node id or adding or removing FSE flow entry or in
SAWF purposes, ieee80211_vif's driver private data is getting
dereferenced by default into ath12k_vif, although VIF belongs
to ath11k. This showed invalid memory reads causing kernel panic
in Direct switching case. Fix this by advertizing ECM
registration support through HW feature flags and reading it
before dereferencing private data.

Patch-depdendency: 953-QSDK-wifi-ath12k-Add-DS-node-id-retrieval-mechanism.patch

Signed-off-by: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/accel_cfg.c | 5 +++++
 drivers/net/wireless/ath/ath12k/fse.c       | 7 +++++++
 drivers/net/wireless/ath/ath12k/mac.c       | 2 ++
 drivers/net/wireless/ath/ath12k/sawf.c      | 5 +++++
 4 files changed, 19 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/accel_cfg.c
+++ b/drivers/net/wireless/ath/ath12k/accel_cfg.c
@@ -28,6 +28,7 @@ static bool ath12k_ds_get_node_id(struct
 	struct ath12k_sta *ahsta;
 	struct ath12k_link_vif *arvif;
 	struct ath12k_base *ab;
+	struct ieee80211_hw *hw;
 
 	wdev = ndev->ieee80211_ptr;
 	if (!wdev) {
@@ -50,6 +51,10 @@ static bool ath12k_ds_get_node_id(struct
 		}
 	}
 
+	hw = wiphy_to_ieee80211_hw(wdev->wiphy);
+	if (!ieee80211_hw_check(hw, SUPPORT_ECM_REGISTRATION))
+		return false;
+
 	vif = wdev_to_ieee80211_vif_vlan(wdev);
 	if (!vif)
 		return false;
--- a/drivers/net/wireless/ath/ath12k/fse.c
+++ b/drivers/net/wireless/ath/ath12k/fse.c
@@ -54,9 +54,16 @@ void *ath12k_fse_get_ab_from_netdev(stru
 	struct ath12k_peer *peer;
 	struct ath12k_vif *ahvif;
 	struct ath12k_link_vif *arvif;
+	struct ieee80211_hw *hw;
 	u8 link_id;
 
 	wdev = dev->ieee80211_ptr;
+	if (!wdev)
+		return NULL;
+
+	hw = wiphy_to_ieee80211_hw(wdev->wiphy);
+	if (!ieee80211_hw_check(hw, SUPPORT_ECM_REGISTRATION))
+		return NULL;
 
 	vif = wdev_to_ieee80211_vif_vlan(wdev);
 	if (!vif)
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -17505,6 +17505,8 @@ static int ath12k_mac_hw_register(struct
 
 	hw->netdev_features |= NETIF_F_LLTX;
 
+	ieee80211_hw_set(hw, SUPPORT_ECM_REGISTRATION);
+
 	ret = ieee80211_register_hw(hw);
 	if (ret) {
 		ath12k_err(NULL, "ieee80211 registration failed: %d\n", ret);
--- a/drivers/net/wireless/ath/ath12k/sawf.c
+++ b/drivers/net/wireless/ath/ath12k/sawf.c
@@ -487,6 +487,7 @@ struct ath12k_base *ath12k_sawf_get_ab_f
 	struct ath12k_link_vif *arvif;
 	struct ieee80211_sta *sta;
 	struct ath12k_sta *ahsta;
+	struct ieee80211_hw *hw;
 	u8 mac_addr[ETH_ALEN] = { 0 };
 	u8 link_id;
 
@@ -497,6 +498,10 @@ struct ath12k_base *ath12k_sawf_get_ab_f
 	if (!wdev)
 		return NULL;
 
+	hw = wiphy_to_ieee80211_hw(wdev->wiphy);
+	if (!ieee80211_hw_check(hw, SUPPORT_ECM_REGISTRATION))
+		return NULL;
+
 	vif = wdev_to_ieee80211_vif(wdev);
 	if (!vif)
 		return NULL;
